ipfscluster_test.go

```diff
@@ -1525,3 +1526,202 @@ func TestClustersGraphUnhealthy(t *testing.T) {
 	}
 	validateClusterGraph(t, graph, clusterIDs)
 }
+
+func TestClustersAddFileLocal(t *testing.T) {
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+
+	// Add the testing directory locally to a random cluster node with
+	// default parameters
+	j := rand.Intn(nClusters)
+	multiFileR, err := test.GetTestingDirMultiReader()
+	if err != nil {
+		t.Fatal(err)
+	}
+	multiPartR := multipart.NewReader(multiFileR, multiFileR.Boundary())
+	params := make(map[string][]string)
+	params["shard"] = []string{"false"}
+	params["quiet"] = []string{"false"}
+	params["silent"] = []string{"false"}
+	params["layout"] = []string{""}
+	params["chunker"] = []string{""}
+	params["raw"] = []string{"false"}
+	params["hidden"] = []string{"false"}
+	params["replMin"] = []string{"-1"}
+	params["replMax"] = []string{"-1"}
+
+	output, err := clusters[j].AddFile(multiPartR, params)
+	if err != nil {
+		t.Error(err)
+	}
+	delay()
+	if len(output) != test.NumTestDirPrints {
+		t.Error("Unexpected number of lines output")
+	}
+	rootCid, _ := cid.Decode(test.TestDirBalancedRootCID)
+
+	fPinnedThunk := func(pinnedCid *cid.Cid, numPins int) func(*testing.T, *Cluster) {
+		return func(t *testing.T, c *Cluster) {
+			v := c.tracker.Status(pinnedCid)
+			if v.Status != api.TrackerStatusPinned {
+				t.Errorf("%s should have been pinned but it is %s",
+					v.Cid,
+					v.Status.String())
+			}
+			allStatus := c.tracker.StatusAll()
+			if len(allStatus) != numPins {
+				t.Errorf("Expected %d pins from local file add but got %d", numPins, len(allStatus))
+			}
+		}
+	}
+	runF(t, clusters, fPinnedThunk(rootCid, 1))
+
+	// Add the same file but as a trickle DAG
+	params["layout"] = []string{"trickle"}
+	multiFileR2, err := test.GetTestingDirMultiReader()
+	if err != nil {
+		t.Fatal(err)
+	}
+	multiPartR2 := multipart.NewReader(multiFileR2, multiFileR2.Boundary())
+	_, err = clusters[j].AddFile(multiPartR2, params)
+	if err != nil {
+		t.Error(err)
+	}
+	delay()
+
+	trickleRoot, _ := cid.Decode(test.TestDirTrickleRootCID)
+	runF(t, clusters, fPinnedThunk(trickleRoot, 2))
+}
+
+func TestClustersAddFileShard(t *testing.T) {
+	clusters, mock := createClusters(t)
+	defer shutdownClusters(t, clusters, mock)
+
+	// Add the testing directory sharded across nodes
+	j := rand.Intn(nClusters)
+	multiFileR, err := test.GetTestingDirMultiReader()
+	if err != nil {
+		t.Fatal(err)
+	}
+	multiPartR := multipart.NewReader(multiFileR, multiFileR.Boundary())
+	params := make(map[string][]string)
+	params["shard"] = []string{"true"}
+	params["quiet"] = []string{"false"}
+	params["silent"] = []string{"false"}
+	params["layout"] = []string{""}
+	params["chunker"] = []string{""}
+	params["raw"] = []string{"false"}
+	params["hidden"] = []string{"false"}
+	params["replMin"] = []string{"-1"}
+	params["replMax"] = []string{"-1"}
+
+	_, err = clusters[j].AddFile(multiPartR, params)
```

Comment:
I probably missed this in some other review. It looks suboptimal that AddFile receives a map of parameters of strings. We should have a type to wrap the parameters with the right types.